### ISSUE TITLE
mkvtoolnix-9.8.0 fails to build with boost-1.63.0

### ISSUE THREAD

**Issue by tgurr** [mbunkus/mkvtoolnix#1886]
I get the following error when trying to build the latest mkvtoolnix:

```
[...]
        LINK src/mkvpropedit
         CXX tests/unit/common/basic_timecode_c.cpp
          CC lib/avilib-0.6.10/avilib.c
src/common/mpeg4_p10.cpp: In function 'memory_cptr mpeg4::p10::parse_sps(const memory_cptr&, mpeg4::p10::sps_info_t&, bool, bool, int64_t)':
src/common/mpeg4_p10.cpp:416:75: error: no matching function for call to 'boost::rational<long int>::assign(unsigned int, long long int)'
       frame_rate.assign(0x80000000, (duration * 0x80000000) / 1000000000ll);
                                                                           ^
compilation terminated due to -Wfatal-errors.
      MSGFMT po/ca.po
/usr/x86_64-pc-linux-gnu/bin/x86_64-pc-linux-gnu-ld: cannot find -lmtxcommon
collect2: error: ld returned 1 exit status
      MSGFMT po/cs.po
      MSGFMT po/de.po
      MSGFMT po/es.po
      MSGFMT po/eu.po
      MSGFMT po/fr.po
      MSGFMT po/it.po
      MSGFMT po/ja.po
         CXX src/input/r_qtmp4.cpp
      MSGFMT po/ko.po
      MSGFMT po/lt.po
      MSGFMT po/nl.po
      MSGFMT po/pl.po
      MSGFMT po/pt.po
      MSGFMT po/pt_BR.po
      MSGFMT po/ru.po
      MSGFMT po/sr_RS.po
      MSGFMT po/sr_RS@latin.po
      MSGFMT po/sv.po
      MSGFMT po/tr.po
      MSGFMT po/uk.po
         UIC src/mkvtoolnix-gui/forms/header_editor/track_type_page.ui
         CXX src/mkvtoolnix-gui/header_editor/track_type_page.cpp
      MSGFMT po/zh_CN.po
      MSGFMT po/zh_TW.po
         CXX src/tools/ebml_validator.cpp
          CC lib/avilib-0.6.10/avimisc.c
         CXX lib/avilib-0.6.10/xio.cpp
         CXX tests/unit/common/bit_cursor.cpp
src/input/r_qtmp4.cpp: In member function 'void qtmp4_demuxer_c::calculate_frame_rate()':
src/input/r_qtmp4.cpp:1936:59: error: no matching function for call to 'boost::rational<long int>::assign(int64_t&, uint32_t&)'
     frame_rate.assign(time_scale, durmap_table[0].duration);
                                                           ^
compilation terminated due to -Wfatal-errors.
[...]
```

I'm not sure if the issue has already been fixed with the latest commits/changes to the mp4 reader.

**Comment by mbunkus** [comment]
I'm developing against Boost 1.63.0 myself, therefore this is somewhat surprising. Which operating system are you compiling on? 32bit or 64bit?

**Comment by mbunkus** [comment]
Additional questions:

* Which compiler are you using? It works fine for me with gcc 6.3.1 and clang 3.9.1.
* How do you run `configure`?
* Do you have any environment variables set like `CFLAGS`, `CXXFLAGS`?

Please try [this patch](https://gist.github.com/mbunkus/6a470e6fb086cae3dbf64839595fb973). I cannot verify whether or not it fixes the issue as I cannot reproduce the problem. The patch does convert all parameters to `long long int`/`int64_t`.

**Comment by tgurr** [comment]
* I'm on 64bit
* Currently using GCC 5.4.0
* configure:
`./configure --build=x86_64-pc-linux-gnu --host=x86_64-pc-linux-gnu --prefix=/usr/x86_64-pc-linux-gnu --bindir=/usr/x86_64-pc-linux-gnu/bin --sbindir=/usr/x86_64-pc-linux-gnu/bin --libdir=/usr/x86_64-pc-linux-gnu/lib --datadir=/usr/share --datarootdir=/usr/share --docdir=/usr/share/doc/mkvtoolnix-9.8.0 --infodir=/usr/share/info --mandir=/usr/share/man --sysconfdir=/etc --localstatedir=/var/lib --disable-dependency-tracking --disable-silent-rules --enable-fast-install --disable-static --with-gettext --with-qt-pkg-config --enable-qt --without-flac`
* CFLAGS, CXXFLAGS:
```
CFLAGS='-pipe -O2 -march=native'
CXXFLAGS='-march=native -O2 -pipe'
```

I'm afraid it still doesn't build for me with your patch applied, now I get:

```
src/input/r_qtmp4.cpp: In member function 'void qtmp4_demuxer_c::calculate_frame_rate()':
src/input/r_qtmp4.cpp:1981:64: error: no matching function for call to 'boost::rational<long int>::assign(long long int, int64_t)'
     frame_rate.assign(1000000000ll, to_nsecs(most_common.first));
                                                                ^
```

**Comment by mbunkus** [comment]
I'm afraid then I cannot help you at the moment. Like I've said: I cannot reproduce this here. If you can come up with a patch that works, I'll gladly accept it.

**Comment by filko** [comment]
As I have the same problem, I looked deeper into this.  If you disable fatal-errors, you get an actual hint from the compiler:

`src/input/r_qtmp4.cpp: In member function 'void qtmp4_demuxer_c::calculate_frame_rate()':
src/input/r_qtmp4.cpp:1949:59: error: no matching function for call to 'boost::rational<long int>::assign(int64_t&, uint32_t&)'
     frame_rate.assign(time_scale, durmap_table[0].duration);
                                                           ^
In file included from src/common/common.h:57:0,
                 from src/common/common_pch.h:4:
/usr/x86_64-pc-linux-gnu/include/boost/rational.hpp:191:13: note: candidate: template<class T> typename boost::enable_if_c<boost::rational_detail::is_compatible_integer<T, IntType>::value, bo
ost::rational<IntType>&>::type boost::rational<IntType>::assign(const T&, const T&) [with T = T; IntType = long int]
     >::type assign(const T& n, const T& d)
             ^
/usr/x86_64-pc-linux-gnu/include/boost/rational.hpp:191:13: note:   template argument deduction/substitution failed:
src/input/r_qtmp4.cpp:1949:59: note:   deduced conflicting types for parameter 'const T' ('long int' and 'uint32_t {aka unsigned int}')
     frame_rate.assign(time_scale, durmap_table[0].duration);
                                                           ^
In file included from src/common/common.h:57:0,
                 from src/common/common_pch.h:4:
/usr/x86_64-pc-linux-gnu/include/boost/rational.hpp:239:13: note: candidate: template<class T> typename boost::enable_if_c<((((std::numeric_limits<T>::is_specialized && std::numeric_limits<T>
::is_integer) && (! boost::rational_detail::is_compatible_integer<T, IntType>::value)) && (std::numeric_limits<T>::radix == std::numeric_limits<_Tp>::radix)) && boost::is_convertible<OtherWra
pped, Iter>::value), boost::rational<IntType>&>::type boost::rational<IntType>::assign(const T&, const T&) [with T = T; IntType = long int]
     >::type assign(const T& n, const T& d)
             ^
/usr/x86_64-pc-linux-gnu/include/boost/rational.hpp:239:13: note:   template argument deduction/substitution failed:
src/input/r_qtmp4.cpp:1949:59: note:   deduced conflicting types for parameter 'const T' ('long int' and 'uint32_t {aka unsigned int}')
     frame_rate.assign(time_scale, durmap_table[0].duration);
                                                           ^`
Basically rational::assign() has 2 overloads both taking homogeneous template arguments.   time_scale is however defined as int64_t and qt_durmap_t::duration is a uint32_t so that won't work.  But that shouldn't work for anyone so it's even more confusing.

**Comment by mbunkus** [comment]
Thank you for looking into it. There are some more facts that don't make sense to me here:

* The patch I've [proposed](https://gist.github.com/mbunkus/6a470e6fb086cae3dbf64839595fb973) casts all arguments to the same type, `int64_t`. However, @tgurr said it doesn't fix the issue.
* Boost's [documentation for v1.63.0](http://www.boost.org/doc/libs/1_63_0/libs/rational/rational.html) doesn't mention a second, overloaded function. It only mentions one.
* In my copy of Boost (Arch Linux, Boost 1.63.0 package revision 1) looking through `/usr/include/boost/rational.hpp` only reveals one definition for `assign`, and one implementation.

**Comment by tgurr** [comment]
I also have some news in the meantime, however they might not be very useful. I had the issue on 2 machines. On one machine I downgraded boost back to 1.61.0 then upgraded back to 1.63.0 and mkvtoolnix started to compile fine.

On my other machine this didn't work, as upgrading to GCC 6.3.0 also didn't help. Here I had to switch my system python interpreter to python 2, recompiled boost 1.63.0 and then compiling mkvtoolnix started to work. Even after switching back my system python interpreter to python 3, recompiling boost 1.63.0 again, compiling mkvtoolnix still works fine.

So right now I'm not able to reproduce the issue. Maybe my system python environment was setup incorrectly to whatever reason and switching it back and forth via our distribution mechanic fixed something which boost 1.63.0 needed for mkvtoolnix to compile correctly(?). I didn't have any issues compiling any other packaged depending on boost.

So in the end I really can't tell what the actual problem was, only a vague guess that it might had something to do with python. I'm not really satisfied without knowing what needed to be fixed to not hit this issue anymore and I'm sorry I probably can't be of any help anymore.

**Comment by mbunkus** [comment]
All very interesting, though not really enlightening, just as you've said. For the record: MKVToolNix' compilation process doesn't use Python at all, and there's no Python code inside MKVToolNix either (meaning it doesn't use any of Boost's Python integration features).

**Comment by mbunkus** [comment]
BTW, which operating system and release are you two using?

**Comment by filko** [comment]
I figured that out.  Our distribution used a wrong boost tarball, I'm not even sure where we got that from.  This ticket should be closed as invalid.

**Comment by mbunkus** [comment]
That sounds plausible :) Thanks for following up on it.

**Comment by tgurr** [comment]
Looks like I've deleted and refetched the tarball then as well while trying to figure out what the problem is on my side. Just for completeness the working tarball seems to be:

```
$ md5sum /var/cache/paludis/distfiles/boost_1_63_0.tar.bz2
1c837ecd990bb022d07e7aab32b09847
```

and the broken one:
```
$ md5sum boost_1_63_0.tar.bz2 
72b85e627cb9dc34dcabb4bac847160c  boost_1_63_0.tar.bz2
```

Looks like Fedora also ran into that issue:
https://lists.fedoraproject.org/archives/list/devel@lists.fedoraproject.org/message/YPUNQB2XBPQRK6OJEHOR3FAQZRQLZTUJ/

And changed their spec file:
https://src.fedoraproject.org/cgit/rpms/boost.git/commit/?id=51c6915fb4d4a613fb19fa33df153af29c853d64

```
-Source0: http://downloads.sourceforge.net/%{name}/%{toplev_dirname}.tar.bz2
+Source0: https://sourceforge.net/projects/boost/files/boost/%{version}/%{toplev_dirname}.tar.bz2
```

And the boost project probably fixed their download in the meantime since while we didn't change our download location from sourceforge now the correct one seems to be downloaded.

So sorry for wasting everyones time (including mine). I wish boost would just have released a new version or at least a fixed tarball with a different name so everything would've been fine. Upstream changing the content of a tarball without renaming it is always causing a huge mess. Should've checked that source of error first though. I apologize.

**Comment by mbunkus** [comment]
No need to apologize! It definitely wasn't your fault. And I wholeheartedly agree that re-releasing a different tarball under the same name is a big no-no. I've been bitten by myself.
